rkt/config/auth.go

```diff
@@ -68,17 +84,90 @@ func (h *basicAuthHeaderer) Header() http.Header {
 	return headers
 }
 
+func (h *basicAuthHeaderer) SignRequest(r *http.Request) *http.Request {
+	r.Header.Set(authHeader, h.GetHeader().Get(authHeader))
+
+	return r
+}
+
 type oAuthBearerTokenHeaderer struct {
 	auth oauthV1
 }
 
-func (h *oAuthBearerTokenHeaderer) Header() http.Header {
+func (h *oAuthBearerTokenHeaderer) GetHeader() http.Header {
 	headers := make(http.Header)
 	headers.Add(authHeader, "Bearer "+h.auth.Token)
 
 	return headers
 }
 
+func (h *oAuthBearerTokenHeaderer) SignRequest(r *http.Request) *http.Request {
+	r.Header.Set(authHeader, h.GetHeader().Get(authHeader))
+
+	return r
+}
+
+type awsAuthHeaderer struct {
+	accessKeyID     string
+	secretAccessKey string
+	region          string
+}
+
+func (h *awsAuthHeaderer) GetHeader() http.Header {
+	return make(http.Header)
+}
+
+func (h *awsAuthHeaderer) SignRequest(r *http.Request) *http.Request {
+	region := h.region
+
+	var body io.ReadSeeker
+	if r.Body != nil {
+		body = r.Body.(io.ReadSeeker)
+	}
+
+	if len(region) == 0 {
+		region = guessAWSRegion(r.URL.Host)
+	}
+	v4.Sign(&request.Request{
+		ClientInfo: metadata.ClientInfo{
+			SigningRegion: region,
+			SigningName:   awsS3Service,
+		},
+		Config: aws.Config{
+			Credentials: credentials.NewStaticCredentials(h.accessKeyID, h.secretAccessKey, ""),
+		},
+		HTTPRequest: r,
+		Body:        body,
+		Time:        time.Now(),
+	})
+
+	return r
+}
+
+func guessAWSRegion(host string) string {
```

Comment:
To be fair, I'm not sure why this small bit of code (with tests) is causing such an uproar. It follows http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region pretty much to the letter. I also don't see what obscure bugs this can cause - it recognizes all documented endpoints, and can cope with new regions as long as they fall into one of the existing categories - if they don't, it is trivial to override the signing region in the config.